extensions/wikia/DesignSystem/scripts/DesignSystemGlobalNavigationNotifications.js

```diff
@@ -0,0 +1,375 @@
+
+				this.$notificationsEntryPoint
+					.mouseenter(this.proxy(this.updateCounts))
+					.mouseenter(this.proxy(this.fetchForCurrentWiki));
+
+				this.$wallNotifications.add($('#pt-wall-notifications'))
+					.on('click', '.notifications-markasread', this.markAllAsReadAllWikis.bind(this));
+
+				$(window).on('resize', $.throttle(50, function () {
+					WallNotifications.setNotificationsHeight();
+				}));
+			},
+
+			checkIfFromMessageBubble: function () {
+				function urlParam(name) {
+					var results = new RegExp('[\\?&]' + name + '=([^&#]*)').exec(window.location.href);
+					if (!results) {
+						return 0;
+					}
+					return results[1] || 0;
+				}
+
+				if (urlParam('showNotifications')) {
+					this.$notifications.trigger('mouseenter');
+					this.$wallNotifications.addClass('show');
+				}
+			},
+
+			updateCounts: function () {
+				this.bucky.timer.start('updateCounts');
+				var data,
+					callback = this.proxy(function (data) {
+						if (data.status !== true || data.html === '') {
+							return;
+						}
+
+						if (data.count) {
+							this.$notifications.removeClass('prehide');
+						}
+
+						this.updateCountsHtml(data);
+
+						// if we already have data for some Wikis, show it
+						this.restoreFromCache();
+
+						// make sure at least 1 element is visible
+						this.showFirst();
+
+						// on first updateCounts show notifications if user
+						// came from Notification bubble on non-wall Wiki
+						this.checkIfFromMessageBubble();
+
+						// do not update for next 10s
+						setTimeout(this.proxy(function () {
+							this.updateInProgress = false;
+						}), 10000);
+
+						this.bucky.timer.stop('updateCounts');
+					});
+
+
+				if (this.updateInProgress === false) {
+					this.updateInProgress = true;
+
+					data = this.getUrlParams();
+
+					nirvana.sendRequest({
+						controller: 'WallNotificationsExternalController',
+						method: 'getUpdateCounts',
+						format: 'json',
+						type: 'GET',
+						data: data,
+						callback: callback
+					});
+				}
+			},
+
+			fetchForCurrentWiki: function () {
+				this.bucky.timer.start('fetchForCurrentWiki');
+				if (this.fetchedCurrent === false) {
+					var wikiEl = this.$wallNotifications.find('.notifications-for-wiki').first(),
+						firstWikiId = parseInt(wikiEl.data('wiki-id'), 10);
+
+					if ( !isNaN(firstWikiId) ) {
+						wikiEl.addClass('show');
+						this.fetchedCurrent = true;
+						this.currentWikiId = firstWikiId;
+						this.wikiShown[firstWikiId] = true;
+						this.updateWiki(firstWikiId);
+						this.bucky.timer.stop('fetchForCurrentWiki');
+					}
+				}
+			},
+
+			restoreFromCache: function () {
+				// after updating counts we update DOM, so all notifications are empty
+				// we update them from cache for all those that still have the same
+				// amount of unread notifications and fetch all the others
+				for (var wikiId in this.notificationsCache) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Restrict what this loop acts on by testing each property. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=javascript%3AForIn)